### PR TITLE
[NEXUS-16978] Tag on publish

### PR DESCRIPTION
**Pull request by bradbeck**

### CODE REVIEW

File: src/test/java/org/sonatype/nexus/ci/util/NxrmUtilTest.groovy [hunk 1]
```diff
@@ -155,20 +159,38 @@ class NxrmUtilTest
...
     } else {
       nxrmConfiguration = new Nxrm3Configuration(id, "internal${id}", 'displayName', 'http://foo.com', 'credentialsId')
     }
 
-    configurationList <<  nxrmConfiguration
-
-    def globalConfiguration = jenkins.getInstance().getDescriptorByType(GlobalNexusConfiguration)
-    globalConfiguration.nxrmConfigs = configurationList
+    globalConfiguration.nxrmConfigs << nxrmConfiguration
     globalConfiguration.save()
```
Comment (jeremybryan) [inline src/test/java/org/sonatype/nexus/ci/util/NxrmUtilTest.groovy]: Nice clean up and conslidation, will boyscout the others to make them consistent.

File: src/main/java/org/sonatype/nexus/ci/nxrm/NexusPublisherBuildStep.groovy [hunk 2]
```diff
@@ -71,20 +86,34 @@ class NexusPublisherBuildStep
...
       NxrmUtil.doCheckNexusRepositoryId(value)
     }
 
+    @Override
     ListBoxModel doFillNexusRepositoryIdItems(@QueryParameter String nexusInstanceId) {
       NxrmUtil.doFillNexusRepositoryIdItems(nexusInstanceId)
     }
+
+    @SuppressWarnings('UnusedMethodParameter')
+    FormValidation doCheckTagName(@QueryParameter String tagName) {
```
Comment (wrightkris) [inline src/main/java/org/sonatype/nexus/ci/nxrm/NexusPublisherBuildStep.groovy]: Q: Do we need these validation methods given we're not actually doing any checks on the tag name (as it's optional)?

File: src/main/java/org/sonatype/nexus/ci/nxrm/v3/ComponentUploaderNxrm3.groovy [hunk 3]
```diff
@@ -34,9 +37,21 @@ import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus3Clien
 class ComponentUploaderNxrm3
     extends ComponentUploader
 {
+  @Override
+  void maybeCreateTag(@Nullable final String tagName) {
+    if (tagName?.trim()) {
+      def nxrmClient = getRepositoryManagerClient(nxrmConfiguration)
+      Optional<Tag> existingTag = nxrmClient.getTag(tagName)
```
Comment (wrightkris) [inline src/main/java/org/sonatype/nexus/ci/nxrm/v3/ComponentUploaderNxrm3.groovy]: trivial: could be trimmed to a 1-line if desired
`nxrmClient.getTag(tagName).orElse(nxrmClient.createTag(tagName))`

File: src/test/java/org/sonatype/nexus/ci/nxrm/ComponentUploaderNxrm3Test.groovy [hunk 4]
```diff
@@ -150,6 +152,61 @@ class ComponentUploaderNxrm3Test
...
+      mockComponentUploader.uploadComponents(publisher, filePath, tagName)
+      tempFile.delete()
+
+    then:
+      1 * client.upload(*_) >> { args ->
+        repo = args[0]
+        component = args[1]
+        tag = args[2]
+      }
+
```
Comment (wrightkris) [inline src/test/java/org/sonatype/nexus/ci/nxrm/ComponentUploaderNxrm3Test.groovy]: Probably should verify the createTag is called as well given the setup:
`if (tagName) {
        1 * client.createTag(tagName)
      }`

not sure if there's a better way to do conditional checks with groovy/spock

**Pull request merged**